standard/variables.md

```diff
@@ -988,3 +998,185 @@ variable_reference
+>
+> ```csharp
+> public class C
+> {
+>
+>     public void M()
+>     {
+>         int[] arr = new int[10];
+>         // element is a reference variable that refers to arr[5]
+>         ref int element = ref arr[5];
+>         element += 5; // arr[5] has been incremented by 5
+>     }     
+> }
+> ```
+>
+> *end example*
+
+A ***reference return*** is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§15.6.1). The variable expression of the reference return is the referent of the reference return.
+
+> *Example:* The following example demonstrates a reference return whose referent is an element of an array field:
+>
+> ```csharp
+> public class C
+> {
+>     private int[] arr = new int[10];
+>
+>     public readonly ref int M()
+>     {
+>         // element is a reference variable that refers to arr[5]
+>         ref int element = ref arr[5];
+>         return ref element; // return reference to arr[5];
+>     }     
+> }
+> ```
+>
+> *end example*
+
+### §ref-safe-contexts Ref safe contexts
+
+All reference variables obey safety rules that ensure the context of the reference variable is not greater than the ref-safe-context of its referent.
+
+For any variable, the ***ref-safe-context*** of that variable is the context where a *variable_reference* (§9.5) to that variable is valid. The referent of a reference variable must have a ref-safe-context that is at least as wide as the context of the reference variable.
+
+> *Note*: The compiler determines the safe context through a static analysis of the program text. The safe context reflects the lifetime of a variable at runtime. *end note*
+
+There are three valid ref-safe-contexts:
+
+- ***declaration-block***: A *variable_reference* to a local variable declared in a block is valid from its declaration to the end of the block in which it is declared. Each nested block represents a different context.  The ref-safe-context of a local variable is the declaration-block in which it is declared. A local variable declared in a method has a ref-safe-context of the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block after the referent, or a nested block.
+- ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter, is valid in the entire function member. The ref-safe-context of the fields of a `struct` type is function-member. A variable with ref-safe-context of function-member is a valid referent only if the reference variable is declared in the same function member.
+- ***caller-context***: Member fields of a `class` type and `ref` parameters have ref-safe-context of caller-context. A variable with ref-safe-context of caller-context can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.
```

Comment:
A comment thread by @BillWagner & @RikkiGibson prompted some archeology in the [bowels of Roslyn](https://github.com/dotnet/roslyn/issues/5233) which uncovered the intended semantics for the context of the implicit `this` (a `ref` parameter) in a `struct` instance method: `this` (and its fields) have function-member context and are **not** safe-to-ref-return like other `ref` parameters.

The following suggestion incorporates these semantics into function-member and caller-context. Also added is text for arrays. The
suggestion includes @BillWagner’s [clarifications](https://github.com/dotnet/csharpstandard/pull/795/files/2976e80dce3b8349533c42521e1eb9c887105594#r1202295671) above.

```suggestion
- ***function-member***: Within a function a *variable_reference* to any of the following has a ref-safe-context of function-member:
  - Value parameters ([§9.2.5](variables.md#925-value-parameters)) on a function member declaration, including the implicit `this` of class member functions; and
  - The implicit reference (`ref`) parameter ([§9.2.6](variables.md#925-reference-parameters)) `this` of a struct member function, along with its fields.

   A *variable_reference* with ref-safe-context of function-member is a valid referent only if the reference variable is declared in the same function member.
- ***caller-context***:  Within a function a *variable_reference* to any of the following has a ref-safe-context of caller-context:
  - Reference (`ref`) parameters ([§9.2.6](variables.md#925-reference-parameters)) other than the implicit `this` of a struct member function;
  - Member fields and elements of such parameters;
  - Member fields of parameters of class type; and
  - Elements of parameters of array type.

  A *variable_reference* with ref-safe-context of caller-context can be the referent of a reference return. A *variable_reference* that can be the referent of a reference-return is ***safe-to-ref-return***.
```

I used nested bullets as the paragraphs were getting too dense. In case the suggestion box does not format well in your browser here it “unboxed”:

- ***function-member***: Within a function a *variable_reference* to any of the following has a ref-safe-context of function-member:
  - Value parameters ([§9.2.5](variables.md#925-value-parameters)) on a function member declaration, including the implicit `this` of class member functions; and
  - The implicit reference (`ref`) parameter ([§9.2.6](variables.md#925-reference-parameters)) `this` of a struct member function, along with its fields.

   A *variable_reference* with ref-safe-context of function-member is a valid referent only if the reference variable is declared in the same function member.
- ***caller-context***:  Within a function a *variable_reference* to any of the following has a ref-safe-context of caller-context:
  - Reference (`ref`) parameters ([§9.2.6](variables.md#925-reference-parameters)) other than the implicit `this` of a struct member function;
  - Member fields and elements of such parameters;
  - Member fields of parameters of class type; and
  - Elements of parameters of array type.

  A *variable_reference* with ref-safe-context of caller-context can be the referent of a reference return. A *variable_reference* that can be the referent of a reference-return is ***safe-to-ref-return***.